### ISSUE TITLE
[WIP] Progress to improve nn.MultiheadAttention and nn.Transformer

### ISSUE THREAD

**Issue by zhangguanheng66** [pytorch/pytorch#32590]
This issue is created to track the progress of the work to improve `nn.MultiheadAttention` and `nn.Transformer`. Since the release of both modules in pytorch v1.2.0, we have received a lot feedbacks from users, including feature requests, bug fix, doc improvement. Here is a quick summary of the work items. Feel free to comment if you have any suggestions.

- [x] 3D attention mask in `nn.MultiheadAttention`

**Comment by zhangguanheng66** [comment]
@cpuhrsch @fmassa @vincentqb

**Comment by snakers4** [comment]
> Add quantization support. Currently, it is blocked by the indexing function here

Hi, I am using built-in PyTorch transformer encoder modules in my custom model (acoustic model + transformer decoder).

In particular I use them like this
```
            layer = nn.TransformerEncoderLayer(d_model=size,
                                                                       nhead=8,
                                                                       dim_feedforward=size * decoder_girth,
                                                                       dropout=dropout)
            self.decoder = nn.TransformerEncoder(layer, decoder_layers)
```
(yeah, I am using encoder layers as a decoder, in my domain this works well!)

Following the new tutorials, I am trying to apply dynamic [quantization](https://pytorch.org/tutorials/intermediate/dynamic_quantization_bert_tutorial.html).

It seems that I have exactly the problem you described

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-100-c8b8d14f108f> in <module>
     12 
     13 inputs = inputs.to(device)
---> 14 model_outputs = quantized_model(inputs)
     15 out0, out = model_outputs
     16 out_cpu = out.cpu().numpy()

/opt/conda/lib/python3.7/site-packages/torch/nn/modules/module.py in __call__(self, *input, **kwargs)
    530             result = self._slow_forward(*input, **kwargs)
    531         else:
--> 532             result = self.forward(*input, **kwargs)
    533         for hook in self._forward_hooks.values():
    534             hook_result = hook(self, input, result)

~/notebook/nvme/aveysov/open_stt_pretrained_deploy/stt_pretrained/models/model.py in forward(self, x)
    113             if self._rnn_type in supported_models:
    114                 x = x.squeeze(1)
--> 115                 x = self.am_encoder_decoder(x)
    116                 x = self.fc(x)
    117                 x = x.transpose(1, 2).transpose(0, 1).contiguous()

/opt/conda/lib/python3.7/site-packages/torch/nn/modules/module.py in __call__(self, *input, **kwargs)
    530             result = self._slow_forward(*input, **kwargs)
    531         else:
--> 532             result = self.forward(*input, **kwargs)
    533         for hook in self._forward_hooks.values():
    534             hook_result = hook(self, input, result)

~/notebook/nvme/aveysov/open_stt_pretrained_deploy/stt_pretrained/models/model.py in forward(self, x)
    340                 # instead of  batch  * channels * length
    341                 return self.decoder(
--> 342                     encoded.permute(2, 0, 1).contiguous()
    343                     ).permute(1, 2, 0).contiguous()
    344             else:

/opt/conda/lib/python3.7/site-packages/torch/nn/modules/module.py in __call__(self, *input, **kwargs)
    530             result = self._slow_forward(*input, **kwargs)
    531         else:
--> 532             result = self.forward(*input, **kwargs)
    533         for hook in self._forward_hooks.values():
    534             hook_result = hook(self, input, result)

/opt/conda/lib/python3.7/site-packages/torch/nn/modules/transformer.py in forward(self, src, mask, src_key_padding_mask)
    174         for i in range(self.num_layers):
    175             output = self.layers[i](output, src_mask=mask,
--> 176                                     src_key_padding_mask=src_key_padding_mask)
    177 
    178         if self.norm:

/opt/conda/lib/python3.7/site-packages/torch/nn/modules/module.py in __call__(self, *input, **kwargs)
    530             result = self._slow_forward(*input, **kwargs)
    531         else:
--> 532             result = self.forward(*input, **kwargs)
    533         for hook in self._forward_hooks.values():
    534             hook_result = hook(self, input, result)

/opt/conda/lib/python3.7/site-packages/torch/nn/modules/transformer.py in forward(self, src, src_mask, src_key_padding_mask)
    281         """
    282         src2 = self.self_attn(src, src, src, attn_mask=src_mask,
--> 283                               key_padding_mask=src_key_padding_mask)[0]
    284         src = src + self.dropout1(src2)
    285         src = self.norm1(src)

/opt/conda/lib/python3.7/site-packages/torch/nn/modules/module.py in __call__(self, *input, **kwargs)
    530             result = self._slow_forward(*input, **kwargs)
    531         else:
--> 532             result = self.forward(*input, **kwargs)
    533         for hook in self._forward_hooks.values():
    534             hook_result = hook(self, input, result)

/opt/conda/lib/python3.7/site-packages/torch/nn/modules/activation.py in forward(self, query, key, value, key_padding_mask, need_weights, attn_mask)
    817                 training=self.training,
    818                 key_padding_mask=key_padding_mask, need_weights=need_weights,
--> 819                 attn_mask=attn_mask)
    820 
    821 

/opt/conda/lib/python3.7/site-packages/torch/nn/functional.py in multi_head_attention_forward(query, key, value, embed_dim_to_check, num_heads, in_proj_weight, in_proj_bias, bias_k, bias_v, add_zero_attn, dropout_p, out_proj_weight, out_proj_bias, training, key_padding_mask, need_weights, attn_mask, use_separate_proj_weight, q_proj_weight, k_proj_weight, v_proj_weight, static_k, static_v)
   3377     assert list(attn_output.size()) == [bsz * num_heads, tgt_len, head_dim]
   3378     attn_output = attn_output.transpose(0, 1).contiguous().view(tgt_len, bsz, embed_dim)
-> 3379     attn_output = linear(attn_output, out_proj_weight, out_proj_bias)
   3380 
   3381     if need_weights:

/opt/conda/lib/python3.7/site-packages/torch/nn/functional.py in linear(input, weight, bias)
   1370         ret = torch.addmm(bias, input, weight.t())
   1371     else:
-> 1372         output = input.matmul(weight.t())
   1373         if bias is not None:
   1374             output += bias

AttributeError: 'function' object has no attribute 't'
```
The problem seems to be with this [function](https://github.com/pytorch/pytorch/blob/ef5637f85e43dff55b03276176ba2949bd83a085/torch/nn/functional.py#L1362-L1383)

Do I understand correctly, that it should go away if this function is replaced by `nn.module` version? Or is there any other way to apply a quick monkey path until `1.5` is released?

**Comment by cpuhrsch** [comment]
Hello @snakers4,

could I kindly ask you to please post a separate issue for this? This issue is related to communicating and planning work around Transformers where you seem to have discovered a concrete bug that we would want to reference separately.

Thanks,
Christian

**Comment by snakers4** [comment]
Hi @cpuhrsch 

> could I kindly ask you to please post a separate issue for this?

https://github.com/pytorch/pytorch/issues/32764

> a concrete bug that we would want to reference separately.

many thanks!

**Comment by mfuntowicz** [comment]
Is there any plan to include a CuDNN (>= 7.5) backend for multihead attentions ? 

https://docs.nvidia.com/deeplearning/sdk/cudnn-release-notes/rel_750.html#rel_750

**Comment by zhangguanheng66** [comment]
> Is there any plan to include a CuDNN (>= 7.5) backend for multihead attentions ?
> 
> https://docs.nvidia.com/deeplearning/sdk/cudnn-release-notes/rel_750.html#rel_750

We had talk with Nvidia before. They are not ready yet. @natalialunova

**Comment by zhangguanheng66** [comment]
Link two issues here. It's not related to nn.MHA but to `float/double` precision.
https://github.com/pytorch/pytorch/issues/34060
https://github.com/pytorch/pytorch/issues/33841

**Comment by zhangguanheng66** [comment]
After decomposing, users should be able to obtain the attention weights before averaging.
https://github.com/pytorch/pytorch/issues/34537

**Comment by zhangguanheng66** [comment]
This is a short review of different attention mechanism used in SOTA research and existing NLP libraries.

AllenNLP

* AdditiveAttention → Computes attention between a vector and a matrix using an additive attention function. The similarity between vector x and matrix y is computed as V tanh(Wx + Uy) [ref](https://arxiv.org/pdf/1409.0473.pdf)
* BilinearAttention → Computes attention between a vector and a matrix using a bilinear attention function. The similarity between x and matrix y is computed as x^TWy + b 
* CosineAttention →Computes attention between a vector and a matrix using a cosine attention function. The cosine similarity between x and matrix y is computed as x/|x| * y/|y|
* DotProductAttention → Computes attention between a vector and a matrix using a dot product function. The dot product similarity between x and matrix y is computed as x * y

From Pytext [link](https://our.internmc.facebook.com/intern/diffusion/FBS/browse/master/fbcode/pytext/models/representations/attention.py)

* DotProductSelfAttention → Computer attention for token vectors x = {t1, t2, ..., t_n}, compute self attention weights to weighs the tokens as softmax(x^T W)
* SequenceAlignedAttention → Given sequences matrix x and y, computes attention weights for each element in x by matching y with each element in P as softmax(relu(x^TW) * relu(y^TW)^T)
* MultiplicativeAttention → Given sequence x and vector y, computes attention weights for each element in P by matching q with each element in P using multiplicative attention as softmax(x * (y^TW))

Research papers:

* Stand-alone self-attention in vision models → given a pixel x(i, j) and extract a local region of pixels (a, b) centered around x(i, j). The single-headed attention for computing the pixel output y(i, j) is sum(softmax(q(i, j)^T k(a,b)) * v(a,b)) where q(i, j), k(a, b), and v(a, b) are the linear project of pixel x(i, j) and pixels (a, b) [ref](https://arxiv.org/pdf/1906.05909.pdf).
* Attention Augmented Convolutional network → very similar to MultiheadAttention. Add the MHA to the convolutional layer as torch.cat(Conv(X), MHA(X)) [ref](https://arxiv.org/pdf/1904.09925.pdf)

**Comment by zhangguanheng66** [comment]
PR to break MHA into building block. Feedback are welcome
https://github.com/pytorch/pytorch/pull/34719

**Comment by hadaev8** [comment]
Also MHA module need to pass bool value in forward and and pytorch checkpointing raise error on it.

**Comment by ghlai9665** [comment]
> * [ ]  The `multi_head_attention_forward` functional is too long to understand. It's also too difficult to modify some parts in the function. The function maintains too many branches. [pytorch/text#720](https://github.com/pytorch/text/pull/720)

@zhangguanheng66 What's the current status of this? I was working with  `multi_head_attention_forward` today and still find it too difficult to understand / modify. Is there anyone working on refactoring it? If not, I'd like to give it a try.

**Comment by zhangguanheng66** [comment]
We refactored the MHA module in torchtext ([link](https://github.com/pytorch/text/blob/master/torchtext/nn/modules/multiheadattention.py)).

**Comment by ghlai9665** [comment]
> We refactored the MHA module in torchtext ([link](https://github.com/pytorch/text/blob/master/torchtext/nn/modules/multiheadattention.py)).

Is there any reason why the current torch transformer doesn't use the new MHA module from torchtext?

**Comment by ghlai9665** [comment]
> > We refactored the MHA module in torchtext ([link](https://github.com/pytorch/text/blob/master/torchtext/nn/modules/multiheadattention.py)).
> 
> Is there any reason why the current torch transformer doesn't use the new MHA module from torchtext?

Just wanted to follow up on this question @zhangguanheng66. I ran to following code to test torchtext's transformer against torch's transformer and the outputs do not match completely. First,  the `torchtext_attn_output` and `torch_attn_output` tensor are only 67% similar. Second, `torchtext_attn_weights` and `torch_attn_weights` have different shapes. Therefore, the Torchtext's transformer seems more than just a refactor. Why am I seeing these inconsistencies and are they deliberate changes?

```
  from torchtext.nn import InProjContainer, MultiheadAttentionContainer, ScaledDotProduct
  from torch.nn.functional import multi_head_attention_forward as mha_forward

  d_model, num_heads, bsz = 10, 5, 64
  query = torch.rand((21, bsz, d_model))
  key = value = torch.rand((16, bsz, d_model))

  torch.manual_seed(42)
  l1 = torch.nn.Linear(d_model, d_model, bias=False)
  l2 = l3 = l4 = copy.deepcopy(l1)
  
  # call torchtext's MHA
  in_proj_container = InProjContainer(l1, l2, l3)
  torchtext_mha = MultiheadAttentionContainer(num_heads, 
                                    in_proj_container, 
                                    ScaledDotProduct(),
                                    l4)
  torchtext_attn_output, torchtext_attn_weights = torchtext_mha(query, key, value)
  
  # call pytorch's MHA
  in_proj_weight = torch.cat([torchtext_mha.in_proj_container.query_proj.weight,
                                  torchtext_mha.in_proj_container.key_proj.weight,
                                  torchtext_mha.in_proj_container.value_proj.weight])
  torch_attn_output, torch_attn_weights = mha_forward(query, key, value,
                          d_model, num_heads,
                          in_proj_weight, None,
                          None, None,
                          False, 0.0,
                          torchtext_mha.out_proj.weight, None)
```

**Comment by zhangguanheng66** [comment]
> > > We refactored the MHA module in torchtext ([link](https://github.com/pytorch/text/blob/master/torchtext/nn/modules/multiheadattention.py)).
> > 
> > 
> > Is there any reason why the current torch transformer doesn't use the new MHA module from torchtext?
> 
> Just wanted to follow up on this question @zhangguanheng66. I ran the following code to test torchtext's transformer against torch's transformer and the outputs do not match completely. First, the `torchtext_attn_output` and `torch_attn_output` tensor are only 67% similar. Second, `torchtext_attn_weights` and `torch_attn_weights` have different shapes. Therefore, the Torchtext's transformer seems more than just a refactor. Why am I seeing these inconsistencies and are they deliberate changes?
> 
> ```
>   from torchtext.nn import InProjContainer, MultiheadAttentionContainer, ScaledDotProduct
>   from torch.nn.functional import multi_head_attention_forward as mha_forward
> 
>   d_model, num_heads, bsz = 10, 5, 64
>   query = torch.rand((21, bsz, d_model))
>   key = value = torch.rand((16, bsz, d_model))
> 
>   torch.manual_seed(42)
>   l1 = torch.nn.Linear(d_model, d_model, bias=False)
>   l2 = l3 = l4 = copy.deepcopy(l1)
>   
>   # call torchtext's MHA
>   in_proj_container = InProjContainer(l1, l2, l3)
>   torchtext_mha = MultiheadAttentionContainer(num_heads, 
>                                     in_proj_container, 
>                                     ScaledDotProduct(),
>                                     l4)
>   torchtext_attn_output, torchtext_attn_weights = torchtext_mha(query, key, value)
>   
>   # call pytorch's MHA
>   in_proj_weight = torch.cat([torchtext_mha.in_proj_container.query_proj.weight,
>                                   torchtext_mha.in_proj_container.key_proj.weight,
>                                   torchtext_mha.in_proj_container.value_proj.weight])
>   torch_attn_output, torch_attn_weights = mha_forward(query, key, value,
>                           d_model, num_heads,
>                           in_proj_weight, None,
>                           None, None,
>                           False, 0.0,
>                           torchtext_mha.out_proj.weight, None)
> ```

We have a CI test to compare torchtext MHA container vs pytorch MHA module ([link](https://github.com/pytorch/text/blob/master/test/data/test_modules.py#L10)). 

For the output attention weights, they are indeed different. In pytorch MHA, it's the average weights over heads and users preferred the non-average weights. Therefore, in the torchtext MHA container, we output the average weights without averaging. And that's why you see the different dimension.

**Comment by ghlai9665** [comment]
> > > > We refactored the MHA module in torchtext ([link](https://github.com/pytorch/text/blob/master/torchtext/nn/modules/multiheadattention.py)).
> > > 
> > > 
> > > Is there any reason why the current torch transformer doesn't use the new MHA module from torchtext?
> > 
> > 
> > Just wanted to follow up on this question @zhangguanheng66. I ran the following code to test torchtext's transformer against torch's transformer and the outputs do not match completely. First, the `torchtext_attn_output` and `torch_attn_output` tensor are only 67% similar. Second, `torchtext_attn_weights` and `torch_attn_weights` have different shapes. Therefore, the Torchtext's transformer seems more than just a refactor. Why am I seeing these inconsistencies and are they deliberate changes?
> > ```
> >   from torchtext.nn import InProjContainer, MultiheadAttentionContainer, ScaledDotProduct
> >   from torch.nn.functional import multi_head_attention_forward as mha_forward
> > 
> >   d_model, num_heads, bsz = 10, 5, 64
> >   query = torch.rand((21, bsz, d_model))
> >   key = value = torch.rand((16, bsz, d_model))
> > 
> >   torch.manual_seed(42)
> >   l1 = torch.nn.Linear(d_model, d_model, bias=False)
> >   l2 = l3 = l4 = copy.deepcopy(l1)
> >   
> >   # call torchtext's MHA
> >   in_proj_container = InProjContainer(l1, l2, l3)
> >   torchtext_mha = MultiheadAttentionContainer(num_heads, 
> >                                     in_proj_container, 
> >                                     ScaledDotProduct(),
> >                                     l4)
> >   torchtext_attn_output, torchtext_attn_weights = torchtext_mha(query, key, value)
> >   
> >   # call pytorch's MHA
> >   in_proj_weight = torch.cat([torchtext_mha.in_proj_container.query_proj.weight,
> >                                   torchtext_mha.in_proj_container.key_proj.weight,
> >                                   torchtext_mha.in_proj_container.value_proj.weight])
> >   torch_attn_output, torch_attn_weights = mha_forward(query, key, value,
> >                           d_model, num_heads,
> >                           in_proj_weight, None,
> >                           None, None,
> >                           False, 0.0,
> >                           torchtext_mha.out_proj.weight, None)
> > ```
> 
> We have a CI test to compare torchtext MHA container vs pytorch MHA module ([link](https://github.com/pytorch/text/blob/master/test/data/test_modules.py#L10)).
> 
> For the output attention weights, they are indeed different. In pytorch MHA, it's the average weights over heads and users preferred the non-average weights. Therefore, in the torchtext MHA container, we output the average weights without averaging. And that's why you see the different dimension.

Yes, I studied that CI test very closely before making the previous post. I understand that the output attention weights have different dimensions but what I am seeing here in my code is different attention output too, which is **unexpected**. Could you double check?

Thanks!

**Comment by isaacmg** [comment]
Another thing that would be useful is an optional parameter to specify the activation function. For instance, in some cases I have wanted to use Entmax or Sparsemax instead of Softmax. But that isn't easily possible with the current setup.  Particularly in [time series many models are entmax in there multiheadattention mechanism](https://proceedings.neurips.cc/paper/2020/file/c6b8c8d762da15fa8dbbdfb6baf9e260-Paper.pdf).

**Comment by jbschlosser** [comment]
FYI for posterity: a flag to disable averaging of attention weights across heads was added in #70055.

**Comment by sgbaird** [comment]
@jbschlosser @bhosmer wondering if the flag for disabling averaging of attention weights can get passed as a `kwarg` to [`nn.TransformerEncoder`](https://pytorch.org/docs/stable/generated/torch.nn.TransformerEncoder.html) (or [`nn.Transformer`](https://pytorch.org/docs/stable/generated/torch.nn.Transformer.html))

**Comment by jbschlosser** [comment]
@sgbaird Currently, `TransformerEncoder` / `TransformerEncoderLayer` don't expose a way to get the self-attention weights at all, averaged or not. For `TransformerEncoder`, note that there would be N sets of attention weights from the N underlying layers. Do you mind opening a new issue requesting this with more details on your use case? Thanks!

**Comment by sgbaird** [comment]
@jbschlosser I opened up some [discussion](https://discuss.pytorch.org/t/how-do-i-obtain-multiple-heads-via-multi-head-attention-forward-when-using-nn-transformerencoder/145047) if you wouldn't mind taking a look. Will make a feature request if that seems appropriate based on what I'm trying to do at a higher level. Thank you!
